isce2_topsapp/iono_proc.py

```diff
@@ -1132,3 +1120,88 @@ def ionSwathBySwath(self: Type[topsApp.TopsInSAR],
     img.filename = outFilename
     img.extraFilename = outFilename + '.vrt'
     img.renderHdr()
+
+
+def ionosphere(self, ionParam):
+
+    ###################################
+    # SET PARAMETERS HERE
+    # THESE SHOULD BE GOOD ENOUGH, NO NEED TO SET IN setup(self)
+    corThresholdAdj = 0.85
+    ###################################
+
+    print('computing ionosphere')
+    # get files
+    unw_filename = self._insar.unwrappedIntFilename
+    cor_filename = self._insar.correlationFilename
+
+    lowerUnwfile = os.path.join(ionParam.ionDirname, ionParam.lowerDirname,
+                                ionParam.mergedDirname, unw_filename)
+    upperUnwfile = os.path.join(ionParam.ionDirname, ionParam.upperDirname,
+                                ionParam.mergedDirname, unw_filename)
+    corfile = os.path.join(ionParam.ionDirname, ionParam.lowerDirname,
+                           ionParam.mergedDirname, cor_filename)
+
+    # use image size from lower unwrapped interferogram
+    img = isceobj.createImage()
+    img.load(lowerUnwfile + '.xml')
+    width = img.width
+    length = img.length
+
+    lowerUnw = np.fromfile(lowerUnwfile, dtype=np.float32)
+    lowerUnw = lowerUnw.reshape(length*2, width)[1:length*2:2, :]
+    upperUnw = np.fromfile(upperUnwfile, dtype=np.float32)
+    upperUnw = upperUnw.reshape(length*2, width)[1:length*2:2, :]
+    lowerAmp = np.fromfile(lowerUnwfile, dtype=np.float32)
+    lowerAmp = lowerAmp.reshape(length*2, width)[0:length*2:2, :]
+    upperAmp = np.fromfile(upperUnwfile, dtype=np.float32)
+    upperAmp = upperAmp.reshape(length*2, width)[0:length*2:2, :]
+    cor = np.fromfile(corfile, dtype=np.float32)
+    cor = cor.reshape(length*2, width)[1:length*2:2, :]
+    amp = np.sqrt(lowerAmp**2+upperAmp**2)
+
+    # masked out user-specified areas
+    if ionParam.maskedAreas is not None:
+        maskedAreas = runIon.reformatMaskedAreas(ionParam.maskedAreas,
+                                                 length, width)
+        for area in maskedAreas:
+            lowerUnw[area[0]:area[1], area[2]:area[3]] = 0
+            upperUnw[area[0]:area[1], area[2]:area[3]] = 0
+            cor[area[0]:area[1], area[2]:area[3]] = 0
+
+    # MG: Check if there is data for compution iono
+    # skip if array is all 0s
+    if np.flatnonzero(lowerUnw).size == 0:
+        print('WARNING: lower/upper band data all zero'
+              ' skip iono computation!')
+        ionos = np.zeros_like(lowerUnw)
+    else:
+        # compute ionosphere
+        fl = runIon.SPEED_OF_LIGHT / ionParam.radarWavelengthLower
+        fu = runIon.SPEED_OF_LIGHT / ionParam.radarWavelengthUpper
+        # polynomial method for removing relative phase unwrapping errors
+        adjFlag = 1
+        ionos = runIon.computeIonosphere(lowerUnw, upperUnw, cor, fl, fu,
+                                         adjFlag, corThresholdAdj, 0)
```

Comment:
I guess this is the important piece that will fix the swath issues.

To be clear, if a swath is all zeros but a swath nearby is not will the ionosphere look ok merged? We probably want to show some examples of Chile here to David.

Reply:
it looks good based on the local run, anyways we do not care about swath over water as that data is not to be trusted anyways. In this case we are going to get additional WARNING that code could not perform swath adjustment. But we can ignore it over water